Sandbox/PongLayer.cpp

```diff
@@ -0,0 +1,235 @@
+#include "PongLayer.h"
+
+#include "Aquarius.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+
+
+static constexpr auto pi = 3.14159265358979323846;
+
+static const struct
+{
+	const float RegularMode = 0.02;
+	const float HyperMode = 0.5;
+} GameMode {};
+
+static auto ballSpeedIncrease = GameMode.RegularMode;
+
+PongLayer::PongLayer()
+	: Layer("Pong", true), m_LeftPaddle(), m_RightPaddle(), m_Ball()
+{
+	app.getEventHandler().subscribe(Aquarius::eventType::KeyPressedEvent,
+		[&](const Aquarius::Event& event)
+		{
+			onEvent(event);
+		});
+}
+
+void PongLayer::onCreation()
+{
+	int height = window.getHeight();
+	int width = window.getWidth();
+
+	glfwSetInputMode(window.get(), GLFW_CURSOR, GLFW_CURSOR_DISABLED);
+
+	m_Camera = std::make_shared<Aquarius::OrthographicCamera>(1, 0.01, height, width);
+	Aquarius::Renderer::Init();
+
+	constexpr auto paddleWidth = 10.0;
+	constexpr auto paddleHeight = 80.0;
+	constexpr auto ballSize = 15.0;
+	constexpr auto leftPaddleSpeed = 0.3;
+	constexpr auto rightPaddleSpeed = 1.0;
+
+	Aquarius::TextureConfiguration textureConfiguration = {
+		Aquarius::TextureWrapOption::Repeat,
+		Aquarius::TextureWrapOption::Repeat,
+		Aquarius::TextureFilteringOption::Nearest,
+		Aquarius::TextureFilteringOption::Nearest,
+	};
+
+	auto paddleTexture = std::make_shared<Aquarius::Texture>("Sandbox/Assets/Paddle.png", textureConfiguration, true);
+	auto ballTexture = std::make_shared<Aquarius::Texture>("Sandbox/Assets/Ball.png", textureConfiguration, true);
+	m_BackgroundTexture = std::make_shared<Aquarius::Texture>("Sandbox/Assets/logo-black-transparent.png", Aquarius::TextureConfiguration{}, true);
+
+	m_LeftPaddle = {
+		paddleTexture,
+		//std::make_unique<MousePaddleController>(),
+		std::make_unique<KeyboardPaddleController>(Aquarius::Input::KeyCode::Key_w, Aquarius::Input::KeyCode::Key_s),
+		//std::make_unique<AIPaddleController>(),
+		{ 20.0, (window.getHeight() / 2.0) - 40.0},
+		{ paddleWidth, paddleHeight },
+		leftPaddleSpeed
+	};
+
+	m_RightPaddle = {
+		paddleTexture,
+		std::make_unique<AIPaddleController>(),
+		{ window.getWidth() - 30.0, (window.getHeight() / 2.0) - 40.0 },
+		{ paddleWidth, paddleHeight },
+		rightPaddleSpeed
+	};
+
+	m_Ball = {
+		{ (window.getWidth() / 2.0) - 5.0, (window.getHeight() / 2.0) - 5.0 },
+		{ ballSize, ballSize },
+		{ 0.4, 0.4 },
+		{ -0.2, 0.0 },
+		this,
+		ballTexture
+	};
+}
+
+void PongLayer::onEvent(const Aquarius::Event& event)
+{
+	if (isActive())
+	{
+		// Static cast instead of dynamic because we know this will be a KeyPressedEvent
+		// no need for a runtime check
+		auto keyPressEvent = static_cast<const Aquarius::KeyPressedEvent&>(event);
+		switch (keyPressEvent.getCode())
+		{
+			case(Aquarius::Input::KeyCode::Key_p):
+			{
+				m_isPaused = !m_isPaused;
+				break;
+			}
+
+			case(Aquarius::Input::KeyCode::Key_h):
```

Comment:
One thing I am wondering is if hyper mode should be activated until the user presses H again. Right now it resets once someone has scored and goes back to normal mode. I do not know if this is possible but it might be something we can refactor for the final submission.